torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,401 @@
+
+    We can also first compose them and then register them. The following code is
+    equivalent to the previous one
+
+        module = nn.Linear(5, 3)
+        # Chain the two parametrizations and register them
+        p2.chain(p1)
+        register_parametrization(module, "weight", p2)
+        # Now p2.evaluate() computes p1 and then p2
+        assert(module.weight == 4. * module.parametrizations.weight.original_tensor())))
+
+
+    Parametrized parameters and buffers have an in-built caching system via the context
+    manager :class:``cached``
+    """
+
+    def evaluate(self):
+        r"""Evaluates the parametrization"""
+        if self.is_registered():
+            return self(self.original)
+        else:
+            raise RuntimeError(
+                "A Parametrization cannot be evaluated before " "registering it"
+            )
+
+    def original_tensor(self):
+        r"""Returns the tensor the parametrization was registered on"""
+        try:
+            return self.last_parametrization().original
+        except AttributeError:
+            raise ValueError("The parametrization is not registered to a tensor.")
+
+    def last_parametrization(self):
+        r"""Returns the last parametrization of the chain.
+
+        In particular, if the parametrization is not chained, it returns itself
+        """
+        last = self
+        while last.is_chained():
+            last = last.parametrizations.original
+        return last
+
+    def chain(self, parametrization):
+        r""" It modifies the current parametrization, parametrizing the original
+        tensor in terms of the new parametrization
+
+        The original parametrization is modified after chain is applied to it
+
+        Args:
+            parametrization (nn.Parametrization): the parametrization to be chained
+        Returns:
+            Module: The original module with the parametrization module appended to it
+        """
+        if not isinstance(parametrization, Parametrization):
+            raise ValueError(
+                "Expecting a Parametrization. Found '{}'".format(type(parametrization))
+            )
+
+        if self.is_registered():
+            raise ValueError(
+                "Cannot chain a parametrization on a parametrization that "
+                "was already registered parametrization."
+            )
+
+        _set_parametrization(self.last_parametrization(), "original", parametrization)
+        return self
+
+    def is_chained(self):
+        r"""Returns True if it is chained to other parametrizations
+        """
+        return is_parametrized(self, "original")
+
+    def is_registered(self):
+        r"""Returns True if it is registered on a module
+        """
+        try:
+            self.original_tensor()
+            return True
+        except ValueError:
+            return False
+
+
+def set_caching(module, tensor_name):
```

Comment:
Oh, that is interesting, why are you doing it like that? I thought that all parametrizations should be cached by default when we enter a `cached()` block